packages/oracle-node/src/fetchers/evm-chain/uniswap-v3-on-chain/UniswapV3FetcherOnChain.ts

```
@@ -0,0 +1,118 @@
+import { DexOnChainFetcher } from "../../dex-on-chain/DexOnChainFetcher";
+import { BigNumber, providers } from "ethers";
+import UniswapV3Quoter from "./UniswapV3Quoter.abi.json";
+import UniswapV3Pool from "./UniswapV3Pool.abi.json";
+import { ethers } from "ethers";
+import { getLastPrice } from "../../../db/local-db";
+import { PoolsConfig, QuoterOutAmount } from "./types";
+
+import { isLiquidity, parseLiquidityDataFeedId } from "../../liquidity/utils";
+
+const DOLLAR_IN_CENTS = 100;
+const DOLLAR_IN_CENTS_EXPONENT = 2;
+const SPOT_PRICE_PRECISION = 10000;
+
+export class UniswapV3FetcherOnChain extends DexOnChainFetcher<QuoterOutAmount> {
+  constructor(
+    name: string,
+    private readonly poolsConfig: PoolsConfig,
+    private readonly provider: providers.JsonRpcProvider
+  ) {
+    super(name);
+  }
+
+  async makeRequest(id: string): Promise<QuoterOutAmount | null> {
+    let assetId = id;
+    if (isLiquidity(id)) {
+      return null;
+    }
+    const { quoterParams, outAssetId, outAssetDecimals, inAssetDecimals } =
+      this.buildParamsForQuoter(
+        this.poolsConfig[assetId].token0Address,
+        this.poolsConfig[assetId].token1Address,
+        assetId
+      );
+
+    const quoterContract = this.getQuoter(assetId);
+    const poolContract = this.getPool(assetId);
+
+    const [amountOutResult, liquidity] = await Promise.all([
+      quoterContract.callStatic.quoteExactInputSingle(quoterParams),
+      poolContract.liquidity(),
+    ]);
+
+    const { amountOut } = amountOutResult;
+    return {
+      amountOut: amountOut as BigNumber,
+      assetId: id,
+      outAssetId,
+      outAssetDecimals,
+      inAssetDecimals,
+      liquidity: liquidity as BigNumber,
+    };
+  }
+
+  override calculateSpotPrice(
+    assetId: string,
+    quoterResponse: QuoterOutAmount
+  ) {
+    const outAssetLastPrice = getLastPrice(quoterResponse.outAssetId)!.value;
```

`outAssetLastPrice` could be undefined here. We should throw an error if so.